homeassistant/config_entries.py

```diff
@@ -859,10 +880,14 @@ def _abort_if_unique_id_configured(
 
         for entry in self._async_current_entries():
             if entry.unique_id == self.unique_id:
-                if updates is not None and not updates.items() <= entry.data.items():
-                    self.hass.config_entries.async_update_entry(
+                if updates is not None:
+                    changed = self.hass.config_entries.async_update_entry(
                         entry, data={**entry.data, **updates}
                     )
+                    if changed and reload_on_update:
+                        self.hass.async_create_task(
+                            self.hass.config_entries.async_reload(entry.entry_id)
```

Comment:
I don't know if we should do this. It's up to the integration to listen to it's own config entry updates in case they want to reload when that happens

Reply:
The `reload_on_update` parameter is there in case one does not want to do that. I think that would be an exceptional case, as the most cases a reload should just work fine (similar to enabling/disabling an entity).

Reply:
Right, but why put it in the core system if a config entry is able to set it up themselves? And they will need to track it anyway since their config entry can be updated from other places too.

Reply:
I think this is more about having less code in general and picking the common default.

So the reason for `_abort_if_unique_id_configured` to trigger, is because the entry obviously already exists. Passing in updated data, reloads. I cannot think of a case where this wouldn't be a valid handling?

If there is one, `_abort_if_unique_id_configured(updates={...}, reload_on_update=False)` can be used and the integration can handle it manually at that point.

Looking at the code base, I could not spot an integration that should not be handled with a reload. Currently `_abort_if_unique_id_configured` is used for updating a config entry with:

- Latest data from discovery (e.g., zeroconf).
- Latest data from manual user input (e.g., when adding the same device manually again as a new device).

Looking at it from the opposite side; when not reloading automatically: Every integration needs to add code for this use case, add a reload method, register/sub that method, ubsub on unload. Which would be a common pattern at that point.

Handling that saves repeated code, adds (IMHO) a sane default behavior, that above all can be disabled completely on integration level.

Reply:
@balloob I agree with you - I think the integration should listen to the update and reload (in fact, this whole thing started with my PR to the Volumio integration that does exactly that).
I do see the point, however, of removing repeated code between integrations, but I think a better approach is to figure out how to abstract away all of the repetitive code in `__init__.py`, with listening code being a small part of it.
In any case, I think it's better left for an arch issue.
My recommendation:
* This PR has a lot of value in avoiding the save, and it should move forward, but without `reload_on_update` (definitely without it defaulting to `True` - the number of tests needing an update should be a warning sign).
* The original Volumio PR (https://github.com/home-assistant/core/pull/38559) moves forward since it's the pattern we currently have in use.
* We open an arch issue to discuss how to simplify boilerplate code for integrations.

Reply:
> the number of tests needing an update should be a warning sign

I'm not sure how that is relevant. The fact that we have 100% coverage on config flows, allows this to be done more safer than anything else. 

Besides, those tests simply didn't mock out `async_setup_entry`. Which is not really related to this feature.

The Volumio pattern PR you are referring to is IMHO the reason to implement this PR at is full. Having all integration implementing what you did in that PR is just nonsense from my perspective.

Reply:
> I'm not sure how that is relevant. The fact that we have 100% coverage on config flows, allows this to be done more safer than anything else.

The problem is not the config flow - you're assuming that reloading the entry doesn't have an effect on those integration. There's about 30 of them, what makes you so sure?

> Having all integration implementing what you did in that PR is just nonsense from my perspective and would create a lot of unneeded listeners.

I think we both agree on the problem, but not on the solution. Listening of updates is just a small subset of a lot of boilerplate code in `__init__.py`. There's forwarding the load, the unload, clearing resources up on `HOMEASSISTANT_STOP` (because unload isn't called on shutdown), creating and populating the data dict, etc...
I think a solution that simplifies all of that, is preferable to a stopgap solution for the specific case of reloading on update. Such a solution could also be opt-in and thus backwards compatible.

Reply:
> There's about 30 of them, what makes you so sure?

Well, we do exactly the same thing when enabling/disabling entities. Which has not raised any issues.

> There's forwarding the load, the unload, clearing resources up on

You can solve problems by stacking them all up in a big pile. 

> Such a solution could also be opt-in and thus backwards compatible.

It is backwards compatible / non-breaking, as said a couple of times now.

Reply:
> Well, we do exactly the same thing when enabling/disabling entities. Which has not raised any issues.

These happen manually (and rarely), not automatically. I don't think it's a fair comparison.

> Lastly, this only effects the config flows, which has 100% test coverage.

That's not true - the entities could already be up when this is happening, and the config flow could be a result of discovery which happens automatically and behind the scenes. If something breaks, the user has no way to understand what happened.

> You can solve problems by stacking them all up in a big pile.

What do you mean? I'm not stacking anything up, just pointing at what I think is the right solution, which is an infrastructure that abstracts away the boilerplate, including listening for updates (which may happen for other reasons, as well). For example, we could have an `IntegrationInitializer` class that you inherit from, which takes care of listening for updates, unloading storing state and passing it on to the platforms, etc... In most cases, the integration only has to define the supported platforms, and initialize the state.

> It is backwards compatible / non-breaking, as said a couple of times now.

That's an assumption. We do not have any guarantee that these integrations are closing all resources properly, etc... They could also be listening for updates anyway, which would cause two reloads.

Reply:
> That's an assumption. We do not have any guarantee that these integrations are closing all resources properly, etc..

Then it was a bug to begin with, which should be fixed.

> They could also be listening for updates anyway, which would cause two reloads.

That isn't an issue.

Reply:
Something like this:
```Python
class VolumioInitializer(IntegrationInitializer):
    def __init__(self):
        super().__init__(PLATFORMS, reload_on_update=True)

    async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
        volumio = Volumio(
            entry.data[CONF_HOST], entry.data[CONF_PORT], async_get_clientsession(hass)
        )
        try:
            info = await volumio.get_system_version()
        except CannotConnectError as error:
            raise ConfigEntryNotReady from error

    return {DATA_VOLUMIO: volumio, DATA_INFO: info}
```

Reply:
and you cal also add another function for closing resources, which the base class will take care of calling both during unload and shutdown.

Reply:
All integrations should unload resources when we call unload entry. Adding more classes to the mix won't help. It's also off topic of this discussion.

The use case of manual user input is not valid. We finish all config flows triggered during async_setup before we set up config entries.

So it would only be for discovery. But if an IP has changed, the config entry wasn't set up anyway as the connection failed, and so we're in "retry_setup" state, and a reload doesn't do anything.

Reply:
> The use case of manual user input is not valid.

In that case, a bit unfortunate, but we have those patterns in our codebase.

>  But if an IP has changed, the config entry wasn't set up anyway as the connection failed, and so we're in "retry_setup" state, and a reload doesn't do anything.

The case here, is the change of the IP (or other data) while HA and the integration is already running. In that case, the setup is already completed, but the device becomes unavailable.

Reply:
I forgot the case where IPs change while HA is already running. This change is ok 👍

Reply:
> They could also be listening for updates anyway, which would cause two reloads.

While it shouldn't cause a problem, it is certainly not desirable behavior so I spent some time and audited every integration using `_abort_if_unique_id_configured` and update listener for this and set the two that had both to `False`.

Reply:
AFAICT this can happen both manually (for example, with Volumio you can get the same UUID when defined manually and through discovery) and automatically at any point. We also cannot assume that the IP is what changed - `updates` can contain any number of other updates, even if the IP hasn't changed.

> Adding more classes to the mix won't help. It's also off topic of this discussion and should not be continued in this thread.

I'm not trying to add more classes. This whole thing was forced on me. All I was trying to do is follow a common pattern, and my PR was blocked in favor of this, without any arch discussion. All I'm saying, is that if we are trying to remove duplication, which was the motivation here, there might be better alternatives, and it's best to discuss in an arch issue.